drivers/lora/sx1276.c

```diff
@@ -112,51 +109,61 @@ void BoardCriticalSectionEnd(uint32_t *mask)
 	irq_unlock(*mask);
 }
 
-uint32_t RtcGetTimerElapsedTime(void)
+u32_t RtcGetTimerValue(void)
 {
-	u32_t ticks;
-	int err;
-
-	err = counter_get_value(dev_data.counter, &ticks);
-	if (err) {
-		LOG_ERR("Failed to read counter value (err %d)", err);
-		return 0;
-	}
+	return k_ms_to_ticks_ceil32(k_uptime_get_32());
+}
 
-	return ticks;
+u32_t RtcGetTimerElapsedTime(void)
+{
+	return (k_ms_to_ticks_ceil32(k_uptime_get_32()) - saved_time);
 }
 
 u32_t RtcGetMinimumTimeout(void)
 {
-	/* TODO: Get this value from counter driver */
-	return 3;
+	return 1;
 }
 
-void RtcSetAlarm(uint32_t timeout)
+void RtcStopAlarm(void)
 {
-	struct counter_alarm_cfg alarm_cfg;
+	k_timer_stop(&dev_data.timer);
+}
 
-	alarm_cfg.flags = 0;
-	alarm_cfg.ticks = timeout;
+static void timer_callback(struct k_timer *_timer)
+{
+	ARG_UNUSED(_timer);
 
-	counter_set_channel_alarm(dev_data.counter, 0, &alarm_cfg);
+	TimerIrqHandler();
 }
 
-uint32_t RtcSetTimerContext(void)
+void RtcSetAlarm(u32_t timeout)
 {
-	return 0;
+	k_timer_start(&dev_data.timer, k_ticks_to_ms_floor32(timeout),
+		      K_NO_WAIT);
 }
 
-uint32_t RtcMs2Tick(uint32_t milliseconds)
+u32_t RtcSetTimerContext(void)
 {
-	return counter_us_to_ticks(dev_data.counter, (milliseconds / 1000));
+	saved_time = k_ms_to_ticks_ceil32(k_uptime_get_32());
```

Comment:
If we want to deal with ticks anyway, why don't use the k_uptime_ticks API instead and store the saved time as an s64_t? The documentation actually states that k_uptime_get_32 uses the full precision clock, so there is normally no benefit to using the 64-bit version.

Reply:
We could most probably get rid of ticks and only deal with real time. The k_timeout_t supports also us or ns timevalues.

Reply:
> This is really just an observation, but couldn't we just define ticks in LoRaMAC-node as being milliseconds and do away with all of the annoying time conversions? That way, tick2ms and ms2tick suddenly become unit operations. We would probably have to return k_ticks_to_ms_ceil32(1) as the minimum timeout in that case though.

The idea is not to change the LoRaMac-node repository. If we start doing this then for sure the delta between upstream and downstream will increase and I don't want to have that :) We should live with what the repository provides. If we want to change something then we _should_ do that in upstream repo and inherit here.

Reply:
> We could most probably get rid of ticks and only deal with real time. The k_timeout_t supports also us or ns timevalues.

Sorry no. It is like abusing the API exposed by LoRaMac-node repo. We already did that once with Rtc/timer but there we had no other choice. But for this case, since we have an API for getting the ticks, I'd like to stick to it.

Reply:
> If we want to change something then we should do that in upstream repo and inherit here.

Maybe post an issue on the Semtech repo and ask if they are receptive to integrating with zephyr?

Reply:
> If we want to deal with ticks anyway, why don't use the k_uptime_ticks API instead and store the saved time as an s64_t? The documentation actually states that k_uptime_get_32 uses the full precision clock, so there is normally no benefit to using the 64-bit version.

Yep, we could do that. Actually it is the right approach. Will do, thanks!

Reply:
> Maybe post an issue on the Semtech repo and ask if they are receptive to integrating with zephyr?

Yeah, we can do that but that is not something which will happen immediately (I mean the change there). So in the meantime, we'll stick with this.

Reply:
It will help with decisions down the road about how much to use the underlying zephyr kernel API's.

Reply:
My thought is Semtech may have an interested to make their stack available across a larger base of devices and reduce their maintenance.

Reply:
> My thought is Semtech may have an interested to make their stack available across a larger base of devices and reduce their maintenance.

But Zephyr is not the only user of their stack...

Reply:
What I meant by defining ticks as ms wouldn't require a change in the LoRaMAC code. The only change it would involve would be the way we expose ticks to the library. It's no different from a real system that just happens to define a tick as 1 ms. This would avoid the weird conversions where we go from microseconds to ticks, and then from ticks to microseconds when scheduling alarms.

If we do that, RtcMs2Tick(x) and RtcTick2Ms(x) would just return x. Calls that take a tick as an argument could just rely on a tick in LoRa being defined as 1ms, or use RtcTick2Ms to document the fact (the call would get inlined).

Reply:
For the record, I've switched to `k_uptime_get_32()` API